Remotion/Web/Development.WebTesting/WebDriver/Factories/BrowserFactoryBase.cs

```diff
@@ -0,0 +1,68 @@
+﻿// This file is part of the re-motion Core Framework (www.re-motion.org)
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+// 
+// The re-motion Core Framework is free software; you can redistribute it 
+// and/or modify it under the terms of the GNU Lesser General Public License 
+// as published by the Free Software Foundation; either version 2.1 of the 
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+// 
+using System;
+using Coypu;
+using Coypu.Drivers;
+using JetBrains.Annotations;
+using OpenQA.Selenium;
+using OpenQA.Selenium.Remote;
+using Remotion.Utilities;
+using Remotion.Web.Development.WebTesting.BrowserSession;
+using Remotion.Web.Development.WebTesting.Configuration;
+
+namespace Remotion.Web.Development.WebTesting.WebDriver.Factories
+{
+  /// <inheritdoc />
+  public abstract class BrowserFactoryBase : IBrowserFactory

```

Comment:
What does the base class do for us?

Reply:
First of all, it eliminates triple code duplication (quadruple if we factor in Firefox) and unifies the class structure of the implementations (little differences in the implementations that achieve the same output make later refactorings much harder). This class may not be ideal, but I do think it's better than before.

Reply:
Which code duplication are we talking about here? The way I see it, there is one bit of shared code (the factory method for the SessionConfiguration) which can be made static and moved somewhere else. The rest of the code actually becomes much harder to read, because it doesn't make any sense without reading both the base class and the respective derived class. Said another way: unless a class only has a few select template methods (which can't or shouldn't be solved via strategies) that also do not already represent the bulk of the business logic, using sub classing isn't a good idea. 

Disclaimer: there can be cases where you want to play with member visibility and syntactical sugar or some technical requirement that mandate the use of sub classing. But those cases are usually never found by simply refactoring code.

Reply:
If I understood you correctly, changing `BrowserFactoryBase` to `BrowserFactory` and injecting the browser-specific differences as strategies instead of relying on subclasses implementing them would **not** fix the issue, since you would still need both classes to fully understand the code, is that right? Although in that particular case, that probably wouldn't be a good idea anyway, since a lot of creation logic would have to be in the strategy instead of the factory.

Reply:
Exactly to both